### PR TITLE
Adding support for ARMv7 on linux

### PR DESCRIPTION
**Pull request by hpux735**
These changes begin support for ARMv7 and linux.  There are a few optimizations yet to be completed (it's falling back to soft floating point), and there seems to be assumptions about the size of ssize made in some swift code. (for example: Foundation/NSXMLParser.swift:238)

### CODE REVIEW

File: lib/target.py [hunk 1]
```diff
@@ -223,8 +311,8 @@ class Vendor:
 
 class Target:
     triple = None
-    sdk = OSType.MacOSX
-    arch = ArchType.x86_64
+    sdk = OSType.UnknownOS
+    arch = ArchType.UnknownArch
     executable_suffix = ""
     dynamic_library_prefix = "lib"
     dynamic_library_suffix = ".dylib"
```
Comment (phausler) [inline lib/target.py]: these are now incorrect if the sdk and arch are defaulted to unknown

File: lib/target.py [hunk 2]
```diff
@@ -51,11 +51,99 @@ class ArchType:
...
     kalimba     = 38
     shave       = 39
-
+# Do not assume that these are 1:1 mapping.  This should follow
+# canonical naming conventions for arm, etc. architectures.
+# See apple/swift PR #608    
+    @staticmethod
+    def to_string(value):
+        if value == ArchType.arm:
+            return "armv7"
```
Comment (phausler) [inline lib/target.py]: isn't armv7a the more common version?

File: lib/target.py [hunk 3]
```diff
@@ -51,11 +51,99 @@ class ArchType:
...
+        return "unknown"
+# Not 1:1, See to_string
     @staticmethod
     def from_string(string):
         if string == "arm":
             return ArchType.arm
+        if string == "armv7":
+            return ArchType.arm
+        if string == "armv7l":
+            return ArchType.arm
```
Comment (phausler) [inline lib/target.py]: i have previously used armv5 armv5tejl armv6 and armv7te; so perhaps this should actually just check a prefix

File: Foundation/NSXMLNodeOptions.swift [hunk 4]
```diff
@@ -46,46 +46,46 @@
 public var NSXMLNodeOptionsNone: Int { return 0 }
 
 // Init
-public var NSXMLNodeIsCDATA: Int { return 1 << 0 }
-public var NSXMLNodeExpandEmptyElement: Int { return 1 << 1 } // <a></a>
-public var NSXMLNodeCompactEmptyElement: Int { return 1 << 2 } // <a/>
-public var NSXMLNodeUseSingleQuotes: Int { return 1 << 3 }
-public var NSXMLNodeUseDoubleQuotes: Int { return 1 << 4 }
-public var NSXMLNodeNeverEscapeContents: Int { return 1 << 5 }
+public var NSXMLNodeIsCDATA: UInt { return 1 << 0 }
```
Comment (phausler) [inline Foundation/NSXMLNodeOptions.swift]: This changes APIs can this be split out to a separate commit? They also present a problem in the regards that UInt does not properly translate to usage patterns in swift since all NSUInteger values are converted to Int not UInt by the importer

File: CoreFoundation/Base.subproj/CFBase.c [hunk 5]
```diff
@@ -875,14 +875,16 @@ void _CFRuntimeSetCFMPresent(void *addr) {
 
 
 extern void __HALT() {
-#if defined(__ppc__)
+#if defined(__ppc__) || defined(__arm__)
```
Comment (phausler) [inline CoreFoundation/Base.subproj/CFBase.c]: actually i think we could probably generalize this into __builtin_trap() since we require clang

File: lib/target.py (hunk 1 above)
Comment (hpux735) [inline lib/target.py]: I think I added the appropriate logic to compensate below.  The only issue I have with setting default values is that it increased the time it took to identify the logic bug that "ArchType.from_string(comps[0])" wasn't actually assigning anything, it was essentially a no-op.  The default value made it look like everything was fine, just wrong.

Comment (phausler) [inline lib/target.py]: fair point; I agree having an opinion about a "default" target is probably incorrect in a general sense. Just was thinking that perhaps it might be good to have these as None or some other thing to avoid any potential other trip-ups

File: lib/target.py [hunk 6]
```diff
@@ -242,20 +328,30 @@ def __init__(self, triple):
...
 
     @staticmethod
     def default():
-        triple = platform.machine() + "-"
+        arch = ArchType.from_string(platform.machine())
+        triple = ArchType.to_string(arch)
         if platform.system() == "Linux":
-            triple += "linux-gnu"
+            if arch == ArchType.arm:
+                triple += "-linux-gnueabihf"
```
Comment (phausler) [inline lib/target.py]: what about "-linux-armeabi" or "-linux-android" etc? I know these are other handled cases in llvm's triple calculations

File: lib/target.py (hunk 1 above)
Comment (hpux735) [inline lib/target.py]: Good call.  I'll do that.

File: lib/target.py (hunk 6 above)
Comment (hpux735) [inline lib/target.py]: Yes, I don't know the best way to handle that.  Do you know a good way to detect that part of the triple would be?

Comment (phausler) [inline lib/target.py]: See http://clang.llvm.org/docs/CrossCompilation.html

arch = x86, arm, thumb, mips, etc.
sub = for ex. on ARM: v5, v6m, v7a, v7m, etc.
vendor = pc, apple, nvidia, ibm, etc.
sys = none, linux, win32, darwin, cuda, etc.
abi = eabi, gnu, android, macho, elf, etc.

So we probably need another field since triples are not three parts and instead five parts with three reasonably optional fields.
arch[sub][-vendor]-sys[-abi]

However this method is the script's host not the target. So I am not certain which hosts we want to really support here.

File: Foundation/NSXMLParser.swift [hunk 7]
```diff
@@ -233,9 +233,9 @@ internal func _NSXMLParserStartElementNs(ctx: _CFXMLInterface, localname: Unsafe
     let parser = ctx.parser
     let reportQNameURI = parser.shouldProcessNamespaces
     let reportNamespaces = parser.shouldReportNamespacePrefixes
-    let prefixLen = prefix == nil ? strlen(UnsafePointer<Int8>(prefix)) : 0
+    let prefixLen = prefix == nil ? UInt(strlen(UnsafePointer<Int8>(prefix))) : 0
```
Comment (phausler) [inline Foundation/NSXMLParser.swift]: isn't the declaration of strlen
`size_t strlen(const char *s)`
which size_t should be unsigned in any reasonable environment typedefed by `__SIZE_TYPE__` hopefully or `unsigned long` if that isn't the case

Which os defines it as `int strlen(const char *s)`?

Comment (hpux735) [inline Foundation/NSXMLParser.swift]: In swift/stdlib/public/SwiftShims/LibcShims.h swift_ssize_t is defined
```
#if defined(__linux__) && defined (__arm__)
typedef      int __swift_ssize_t;
#else
 typedef long int __swift_ssize_t;
#endif
```

 and later in swift/stdlib/public/stubs/LibcShims.cpp it is checked with a static assert to ensure that swift_ssize_t matches whatever the C ssize_t is.  Because of this, the size of ssize_t cannot be assumed to be a long int.

```
static_assert(std::is_same<ssize_t, swift::__swift_ssize_t>::value,
              "__swift_ssize_t is wrong");
```

I definitely don't think that this commit is the right way to do it.  This is another case of trying to get it to finish building to check linking.  I can revert this to what's existing for the purposes of this pull request.

Comment (phausler) [inline Foundation/NSXMLParser.swift]: nah; if that is the current definition then lets roll with it and just submit a bug report on bugs.swift.org with perhaps a comment in the top of this file to explain why that is the case.

Comment (hpux735) [inline Foundation/NSXMLParser.swift]: Done: https://bugs.swift.org/browse/SR-314

**Comment by phausler** [comment]
during the build it seems that it is spewing out a warning
```
[174/176] CompileSwift: Foundation/NSURLError.swift
<unknown>:0: warning: argument unused during compilation: '-mcpu=cortex-a15'
```
Is this because I am missing some other commit from the compiler?

**Comment by hpux735** [comment]
There is a compiler issue at play there, but it isn't resolved yet.  I backed out.

**Comment by phausler** [comment]
That warning was actually from a x86_64 host building for a x86_64 target; ubuntu 14

**Comment by hpux735** [comment]
It's nice that the compiler was smart enough to ignore obviously contradictory information! :)

**Comment by phausler** [comment]
full clean build for the entire toolchain + foundation seems to pass just fine on ubuntu and on mac os x; the only caveat here is that this falls under the same domain as the FreeBSD port in that this is not a regularly tested path so there may be things that break since we won't be testing there.

**Comment by hpux735** [comment]
Yep, that's totally expected.  Thanks!!

File: Foundation/NSXMLParser.swift (hunk 7 above)
Comment (hpux735) [inline Foundation/NSXMLParser.swift]: Apparently this is intended behavior.  Jordan Rose closed the bug report.

Comment (gribozavr) [inline Foundation/NSXMLParser.swift]: `strlen()` is returning `size_t`, not `ssize_t` (note the double s).  I think this issue is worth further investigation.

Comment (hpux735) [inline Foundation/NSXMLParser.swift]: Although it's true that size_t (and swift_size_t) is distinct from swift_ssize_t it still changes size based upon the underlying architecture.  It is ultimately typedef'd from __SIZE_TYPE__ which, although I couldn't find its actual definition) is 8 bytes on x86_64 (linux or mac) and 4 bytes on 32-bit arm.

I still don't think this is the best solution in this particular instance.  The error we're addressing here is:

error: cannot convert value of type 'UInt32' to expected argument type 'UInt'

So, it is my expectation that we're promoting a 32-bit unsigned into to a 64-bit unsigned int, or simply removing the bit-size specialization, with it remaining 32-bits.  Either way, it's safe in the 32-bit case.  In 64-bit code, this should be a no-op, because it's a cast from UInt to UInt.

Comment (gribozavr) [inline Foundation/NSXMLParser.swift]: `size_t` is supposed to be imported as `Int` in Swift, the importer has a special case for it.  Would you mind checking if there is an issue with that on Linux/armv7?

```Swift
import Glibc
print(size_t.self) // should print 'Int'
```

Comment (hpux735) [inline Foundation/NSXMLParser.swift]: Yes,  It is.

```
wdillon@tegra-ubuntu:~$ cat size_t.swift 
import Glibc
print(size_t.self) // should print 'Int'

wdillon@tegra-ubuntu:~$ swiftc size_t.swift 
wdillon@tegra-ubuntu:~$ ./size_t 
Int
wdillon@tegra-ubuntu:~$ uname -s -m
Linux armv7l
```

At some pint strlen's return value becomes Uint32.
